### PR TITLE
texlive: transform into overridable scope and extract generic interface

### PR DESCRIPTION
**Pull request by apfelkuchen6**
## Description of changes

This includes all the preparation neccessary to add texlive-2023 ( #236382 ) and builds upon #250621 and #248746.

Unfortunately I wasn't really able to divide this into smaller independent chunks that can be merged separately, but I think it should be possible to comprehend the changes when looking at them commit-by-commit.

This should also not cause any rebuilds.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by xworld21** [comment]
Before jumping into the details, the treewide bit is problematic – better to mark this PR as draft for the time being? I'd rather do a single treewide switch to `buildEnv` before 23.11, basically do only one user visible transition, unless we discover it's not materially feasible. In the meantime I'll do my best to understand what's going on with these splices, see how it plays with the multi-output changes.

**Comment by apfelkuchen6** [comment]
I fully agree that we should only do one treewide change once all the changes are in place. This treewide change should be done in a separate PR somewhen in the future.

The "treewide" commit in this PR just changes the things with mutual texlive dependencies (asymptote, biber, biber-ms) to make `biber = biber.override { texlive = self; }` work (self doesn't have the compat-fixups applied yet). I can drop the changes to these programs and remove the overrides for now. As long as there is only one texlive, there should not be any problems.

File: pkgs/tools/typesetting/tex/texlive/default.nix [hunk 1]
```diff
@@ -154,15 +141,36 @@ in
...
+  compatFixups = scope:
+    # TODO
+    scope.pkgs // # remove this line to fix cross
+    scope // {
+      bin = scope.bin // {
+        latexindent = lib.findFirst (p: p.tlType == "bin") scope.pkgs.latexindent.pkgs;
+      };
+    };
+
+in applyOverScope compatFixups texlive
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/default.nix]: Now that I understand what is going on, I managed to verify that
```nix
lib.last pkgsCross.aarch64-multiplatform.__splicedPackages.texlive.pkgs.a2ping.pkgs
```
does not have its own splices. I assume that the binary containers need to be spliced as well, don't they? So we really need the multi-output to make this work properly.

Comment (apfelkuchen6) [reply pkgs/tools/typesetting/tex/texlive/default.nix]: Yes, ideally, the bin containers should also be spliced. This isn't a huge deal yet as there isn't anything cross-compilable that uses the bin containers yet and I doubt users will appear while the interface is this terrible and they aren't really exposed.

But this is probably™ something that fixes itself once the better format for the containers lands.

File: pkgs/tools/typesetting/tex/texlive/default.nix [hunk 2]
```diff
@@ -2,79 +2,20 @@
...
-, python3, ruby, perl, tk, jdk, bash, snobol4
-, coreutils, findutils, gawk, getopt, gnugrep, gnumake, gnupg, gnused, gzip, ncurses, zip
-, libfaketime, asymptote, biber-ms, makeFontsConf
-, useFixedHashes ? true
+{ lib
+, makeScopeWithSplicing', pkgsBuildBuild, pkgsBuildHost, pkgsBuildTarget, pkgsHostHost, pkgsHostTarget
 , recurseIntoAttrs
+, fetchurl, runCommand
+, ghostscript_headless, harfbuzz
+, tlpdb, tlpdbxzHash, src, version, mirrors, useFixedHashes ? true, fixedHashes ? {}
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/default.nix:10]: This line is crucial. It took me a while to understand that any flat TeX Live package scope cannot contain its own tlpdb, because the operator `//` is not lazy enough to allow for that kind of recursion. Hence `texlive` must be a function of TLPDB, or better of the TeX Live repository: the root of the repo (`mirrors`), the hash of texlive.tlpdb.xz (`tlpdbxzHash`), the nixified tlpdb (`tlpdb`), optionally the source of the binaries (`src`), optionally fixed hashes, and some metadata e.g. `version` but also say the snapshot date.

What's missing here is support for other repos, which we should support eventually, for tlcontrib. So we need a suitable generic interface. Best I can suggest, `texlive` should take an ordered list of respositories which are merged in a single package scope, and each repo is an attribute set containing the above `tlpdb`, `hash`, `src`, `mirrors`, and e.g. `extraVersion`, `name` (not sure about `fixedHashes` yet).

Comment (apfelkuchen6) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: I don't really like the interface you suggest, but we definitely should think about how to handle tlcontrib now.

Just taking a `repos`-list parameter makes overriding incredibly clumsy, and I don't think we colud possibly need the full power of that anyway -- There are no other third party repositories. The binary sources should be kept separate as they don't really belong to one of the repositories.

How about separate `src` and `binversion` attributes and an attrset `repos`, where each attribute is itself an attribute with `tlpdb` and `mirrors` and optional `version`, `hash` and `fixedHashes` attributes? One small downside is that it's not as obvious in which order the nixified `tlpdb`s are folded in case of conflicts. Alternatively, I'd also like naming the attribute `extraRepos` and keeping the attributes for texlive at the toplevel to keep overriding easy.

We could combine the version numbers into something like `2023+texlive-20230828+tlcontrib-20230828`, which sounds incredibly awkward, but is accurate!

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: Agree, repo list is not a great idea. Let's remember that `texlive` must have a base repo which is an actual TeX Live, binaries included, and the binaries are coupled with the repo (run, doc, tlpkg must match - so one can override the repo part, but not by that much).

Given that, we can only stack repos (tlcontrib) on top of the base with no new binaries. Hence the starting point should be your code, and a passthru function `withRepos`? With an implicit override behaviour, from left to right, each new repo can overwrite a previous package. And there could be an attribute `extraRepos` too from the start.

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: (Further thought: `tlpdbxz` can be a single `fetchurl`, making `tlpdbxzHash` and `mirrors` redundant. And we could put `binversion` in the `passthru` of the source `fetchurl`, saving another argument.)

Comment (apfelkuchen6) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: > (Further thought: tlpdbxz can be a single fetchurl, making tlpdbxzHash and mirrors redundant. And we could put binversion in the passthru of the source fetchurl, saving another argument.)

I think extracting the mirror from the `tlpdbxz` derivation is somehow unexpected (and explodes if you override it with a local file), so I'd rather pass the mirror in separately even though extracting the url from the derivation should usually do the right thing and save an argument. Minimizing the number of arguments shouldn't isn't really a goal, having a simple and intuitive interface is. Piggy-backing optional stuff such as version numbers onto the derivations is fine though.

> Hence the starting point should be your code, and a passthru function withRepos? With an implicit override behaviour, from left to right, each new repo can overwrite a previous package. And there could be an attribute extraRepos too from the start.

Do I understand correctly that you want the interface to be something like
```
texlive.combined.scheme-minimal.withRepos ( { ... } ).withPackages ( p: [ ... ] )
```
?
This sounds nice for the user, but a little annoying to implement: if we want `withRepos` to essentially `texlive.override { extraRepos = ... }`, it needs “outside access“ to texlive as the override is not available inside the scope. This means that it can be only added as a kind of wrapper and we need to make sure that every externally accessible way to produce a texlive environment adds this modifier to ensure a consistent user interface. At the same time, internal code cannot use these modifiers as they are not applied yet. It should be possible, but it probably involves a lot of spaghetti wrapper code, especially if combined with the other modifiers.

Just allowing to use `.override` before combining would be a lot simpler.

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: Concrete proposal: every repository is represented by `tlpdb = fetchurl { urls = [ ... ]; hash = "..."; passthru.nix = import ./tlpdb.nix; passthru.fixedHashes = { }; }` (could be named `tlpdbTeXLive`, `tlpdbContrib`). This contains all the information about the repo and is linked to its precomputed nix expression, plus it enables easy local updates with import-from-derivation.

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: > I think extracting the mirror from the `tlpdbxz` derivation is somehow unexpected (and explodes if you override it with a local file), so I'd rather pass the mirror in separately even though extracting the url from the derivation should usually do the right thing and save an argument.

What would be the use case? `texlive.tlpdb.xz` is an index of a TeX Live archive which must live in the same tree as the packages, if it is local, the assumption is that the packages are local as well (e.g. that's how `tlmgr` works).

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: (I wrote the "concrete proposal" before I got your other message!)

Comment (apfelkuchen6) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: > What would be the use case? texlive.tlpdb.xz is an index of a TeX Live archive which must live in the same tree as the packages, if it is local, the assumption is that the packages are local as well (e.g. that's how tlmgr works).

The use case I had in mind is "I want to change a lot of stuff and think changing the tlpdb is easier than fiddling around with the tlpdb overrides", but "a lot of stuff" is severely limited if one cannot change the actual package sources that are referenced. So not supporting this after all might be fine after all. Or checking `tlpdb.passthru.urls` first and only using `tlpdb.urls` if it doesn't exist to allow the user to supply any `tlpdbxz`/`url` combination, regardless of whether that makes sense or not.

The more I think about it, the more I like your proposal :+1:

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: > if we want `withRepos` to essentially `texlive.override { extraRepos = ... }`, it needs “outside access“ to texlive as the override is not available inside the scope.

I guess in my mind there should be a function `makeTeXScope` which takes a repo and generate the attribute set, which is easy to make overrideable in this way. In practice I think `withRepos` would fit into the `applyOverScope` wrapper so we don't need additional code.

The advantage of the `withPackages`, `withRepos` aliases is that they can be made composable. At least I have written `withPackages` so that it adds packages instead of overriding them. But that's an implementation detail, first we have to agree on what a repo is!

> Or checking `tlpdb.passthru.urls` first and only using `tlpdb.urls` if it doesn't exist to allow the user to supply any `tlpdbxz`/`url` combination,

Right, let's take my proposal as a default behaviour but with an obvious override (`passthru.mirrors` though, lest we encourage people to break `find-tarballs.nix`!).

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: ...and after reviewing and simplifying `core`/`core-big`, I am circling back to the idea of `texlive-bin`. Let's pretend we pull back `core-big`, `xindy`, `xpdfopen` etc. under a single derivation. Let's also pretend that we hardcode the new `binToOutput` (I'll probably do this part anyway, to make upgrades more understandable). Now the binaries are a single derivation that is virtually independent of `tlpdbxz`, although it cannot be used on its own.

With this in mind, top level `texliveBin`, `texlive` would allow e.g.
```nix
{ texlive2023Bin = texliveBin.overrideAttrs (_: { src = /*...*/; version = "2023"; /*...*/; });
  texlive2023 = texlive.override { texliveBin = texlive2023Bin; tlpdbxz = /*...*/; /*...*/; }; }
```
This feels a lot less awkward than overriding `src` and attaching the TeX Live year out of band. Also less awkward than trying to make a scope with binaries and packages in the same tree. An extreme use case: if I want to run a bleeding edge SVN snapshot, I need to override both the binary sources *and* whatever patches are being applied; with `texliveBin`, people should be able to do it without special documentation.

The only oddity is that `texliveBin` cannot be used directly, but we could just set `meta.outputsToInstall = []` like I suggested for TeX Live packages and put a warning in the description. The packages would still form a scope, but without special attributes for binaries.

TL;DR top level `texliveBin` and `texlive` seem to be a lot simpler to implement and use if we put `bin.core` back together, compared to the other ideas we have been trying?

PS: actually, we can make `texliveBin` work correctly even with the `core`/`core-big`/etc split – I just couldn't see that before getting my hands dirty. We could use `mkDerivation (finalAttrs: ...)` and `passthru.xindy`, `passthru.xpdfopen`, `passthru.extra` so that overrides trickle down correctly. Still awkward, but maybe less awkward than the schemes we have cooked up so far.

Comment (apfelkuchen6) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: > Now the binaries are a single derivation that is virtually independent of tlpdbxz, although it cannot be used on its own.

But we still have the annoying xindy -> texlive.combined.scheme-small dependency thath needs to be dealt with somehow. I guess we could hardcode the dependency (i.e. also use latex from texlive-2022 to build xindy from texlive-2023, but that's really ugly) or try to move xindy directly to the bin containers and not build it in texliveBin (ugh).

> With this in mind, top level texliveBin, texlive would allow e.g.
>
> ```
> { texlive2023Bin = texliveBin.overrideAttrs (_: { src = /*...*/; version = "2023"; /*...*/; });
>   texlive2023 = texlive.override { texliveBin = texlive2023Bin; tlpdbxz = /*...*/; /*...*/; }; }
> ```

I'm not really sure.

I think having a single "real" top level attribute `texliveBin` that could be used as `((texliveBin.override { src = ..; tlpdb = ...; packageOverrides = ...;})).overrideAttrs (old: { patches = old.patches ++ [ ... ]; });.withPackages (p: [ p.scheme-full ])` and a buch of convience aliases is less confusing and more convenient than separating the package set from the package set completely and managing them seperately. And as it is exactly what the other languages do, it's hopefully not be that hard to implement (well, modulo the annoying `xindy` problem).

> PS: actually, we can make texliveBin work correctly even with the core/core-big/etc split – I just couldn't see that before getting my hands dirty. We could use mkDerivation (finalAttrs: ...) and passthru.xindy, passthru.xpdfopen, passthru.extra so that overrides trickle down correctly. Still awkward, but maybe less awkward than the schemes we have cooked up so far.

I'd rather avoid having half of the binaries in split outputs and the other half attached as passthru attributes. I don't really see the benefit of the separate derivations anyway, so I think Just merging everything (except xindy...) is better.

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: > I think having a single "real" top level attribute `texliveBin`

Yes, I agree actually. I had in mind `texlive` as scheme-infraonly, but that just seems to complicate things. So `texlive` as (uninstallable) binaries and `withPackages` on top?

I would still prefer for the overrides to look identical to you override a normal derivation, so e.g. we should be using
```
texlive.overrideAttrs (_: { src = /*...*/; version = "2020"; })
```
In practice, I would prefer `texlive2023 = callPackage ./.../2023.nix { };` with `2023.nix` written using `overrideAttrs`. Not quite the same as the other interpreters, but I would say that it is a good thing here.

For `tlpdbxz`, I am tempted to say it should be `texlive.passthru.tlpdbxz` and overrideable via `overrideAttrs`, at least until we figure where to fit tlcontrib? Mostly to avoid the `stable.nix` wrapper with its own `override` thing.

At any rate, you are convincing me that `texlive` should just be the binaries with lots of `passthru` – the content of which needs to be spelled out. In the meanwhile, that's enough to finish `buildEnv` and `withPackages`.

Re xindy, which I keep forgetting about (darwin!), we must keep building it separately apparently. I gather it needs a full scheme incl. formats to build the inputenc module so we can't avoid the circular dependency. We could do the equivalent of
```nix
xindy = texliveBinaries.overrideAttrs (prev: {
  configureFlags = prev.configureFlags ++ [ "--disable-all-pkgs" "--enable-xindy" ];
  nativeBuildInputs = prev.nativeBuildInputs ++ [ scheme-basic.withPackages/*...*/ ];
}
```
within the packages scope, so that `src` and other build configs are consistent in face of overrides.

Comment (apfelkuchen6) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: > Yes, I agree actually. I had in mind texlive as scheme-infraonly, but that just seems to complicate things. So texlive as (uninstallable) binaries and withPackages on top?

Do you know whether it's possible to make it actually uninstallable (i.e. fail eval).  Setting `outputsToInstall = []` just makes installing it a noop, right? I'm slightly worried about users being frustrated when installing `texlive` doesn't work, but I'm probably overthinking this. Adding a helpful description is probably enough for most users.

But yes, this structure sounds like the most sensible to me. We could even attach `tests.texlive` as `passthru.tests` to `texlive` to have ofborg automagically build the tests when a commit starts with `texlive: `.

> `texlive.overrideAttrs (_: { src = /*...*/; version = "2020"; })`

This is definitely better than having to use `.override`. Also using `.overrideAttrs` for tlpdb also sounds logical. We should definitely make sure that the mechanism we choose for the package overrides is sufficiciently composable such that when a users uses it on texlive-2023, they don't remove the previous 2023-specific overrides (so we probably don't want to just copy python, see #44426).

> ```
> xindy = texliveBinaries.overrideAttrs (prev: {
>  configureFlags = prev.configureFlags ++ [ "--disable-all-pkgs" "--enable-xindy" ];
>  nativeBuildInputs = prev.nativeBuildInputs ++ [ scheme-basic.withPackages/*...*/ ];>
> }
> ```

That looks like a neat solution. Hopefully later configure flags override the behaviour of the former ones.

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: Ok, we are converging on something sensible.

> Do you know whether it's possible to make it actually uninstallable (i.e. fail eval). Setting `outputsToInstall = []` just makes installing it a noop, right?

If we set `met.outputsToInstall = [ "doNotInstall" ]; passthru.doNotInstall = throw "..."`, then `buildEnv` will fail. Not sure how `nix-env` will take it though.

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: > previous 2023-specific overrides (so we probably don't want to just copy python, see #44426)

Re `packageOverrides`, I'd rather omit it altogether exactly because of that problem. `withPackages` is more obviously more composable (incidentally, it does *not* compose for Python). Presumably we'd have `withConfig` (not the final name) which does the same for other overrides, e.g. `extraOutputsToInstall` gets concatenated, `paperSize` goes through `//`, etc.

So far we seem safe: `texlive.override` changes the inputs, same as rest of Nixpkgs; `texlive.overrideAttrs` changes the binaries and/or `tlpdbxz`, is composable by construction via `prev`, inherited by `xindy` et al. if we play it right; `texlive.pkgs.overrideScope` (possibly lifted/moved to `texlive.overrideScope`) also inherently composable. Then auxiliary methods `withPackages`, `withConfig`, with composability built in.

Comment (apfelkuchen6) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: The difficult part is making sure that `texlive.overrideScope (self: super: { ... }).overrideAttrs {...}` does not lose the package overrides. I don't think the naive approach "just works", but it can probably made to work by entangling the different override functions in an interesting manner (similar to `applyOverScope` here).

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: > I don't think the naive approach "just works"

Uhm, I think the only possible implementation is
```nix
passthru.overrideScope = f: finalAttrs.finalPackage.overrideAttrs (prev: {
  passthru = prev.passthru // {
    pkgs = prev.passthru.pkgs.overrideScope f;
  };
};
```
and that should compose as correctly as `overrideAttrs` itself. I think. If it's not infinite recursion.

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: Ah, there's another piece of design I need to complete `buildEnv`. We are converging on:
- `.override` to change the inputs (e.g. `ghostscript = ghostscriptX`, or `enableStatic = true`); standard `callPackage` but possibly extra wrappers to allow `texlive.combine.scheme-small.override`
- `.overrideAttrs` to modify the `core` derivation, `passthru.tlpdbxz` and similar; based on `mkDerivation (finalAttrs: _)`
- `.overrideScope` to modify `texlive.pkgs` while respecting interdependencies; carefully lifted to `texlive` for convenience
- `.withPackages` to create/add packages to a combined environment

Presumably all of the above methods can be called anywhere, in any order (even `.withPackages (_)).override`?).

But now we have the config arguments of `buildEnv`: `extraOutputsToInstall`, `paperSize`. Some options:
1. `.override`: pass them as global argument (no easy composition)
2. `.overrideAttrs`: set them in `passthru` (ugly but composable)
3. local `.override`: just replace the top `.override` – this is the Python solution `buildEnv.override` (same composability problems as 1.)
4. `.overrideConfig` / `.withConfig`: one replaces the config, one *adds* config, e.g. *more* outputs; not exposed at top level, just in combined schemes, for clarity (good composability via `.overrideConfig (prev: ...)`, of which `.withConfig` is just a special case)
5. `.overrideEnv`: like `.overrideConfig` but emphasis on what receives the override (but no `.withEnv` because it is unreadable)

Any other approaches? Now that I listed them, I prefer 4. over the others, but maybe only because it's more fleshed out.

We might still use 2. internally, so that it composes correctly for free, and write `.overrideConfig` the same way we lift `.overrideScope`.

(Side note: we have to split `core` anyway because we can't quite split the libraries, and we need a small `dev` output for Kpathsea and synctex. Otherwise we also get the LuaTeX library, which is a few MBs.)

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:10]: Quick update on a very long conversation: I have a proof-of-concept implementation of `texlive` as a big `mkDerivation (finalAttrs: ...)`, with `bin.core`, `pkgs`, `tlpdb` etc. as `passthru` attributes. It actually overrides pretty well, enough that changing the TeX Live year becomes as easy as changing a hash when allowing recursive nix. More on this in one or two weeks.
